test/integration/util/kube_apiserver/apiserver_test.go

```diff
@@ -159,3 +161,144 @@ func (suite *testSuite) TestKubeEvents() {
 	assert.Len(suite.T(), added, 2)
 	assert.Len(suite.T(), modified, 0)
 }
+
+func (suite *testSuite) TestServiceMapper() {
+	c, err := apiserver.GetCoreV1Client()
+	require.Nil(suite.T(), err)
+
+	// Create a Ready Schedulable node
+	// As we don't have a controller they don't need to have some heartbeat mechanism
+	node := &v1.Node{
+		Spec: v1.NodeSpec{
+			PodCIDR:       "192.168.1.0/24",
+			Unschedulable: false,
+		},
+		Status: v1.NodeStatus{
+			Addresses: []v1.NodeAddress{
+				{
+					Address: "172.31.119.125",
+					Type:    "InternalIP",
+				},
+				{
+					Address: "ip-172-31-119-125.eu-west-1.compute.internal",
+					Type:    "InternalDNS",
+				},
+				{
+					Address: "ip-172-31-119-125.eu-west-1.compute.internal",
+					Type:    "Hostname",
+				},
+			},
+			Conditions: []v1.NodeCondition{
+				{
+					Type:    "Ready",
+					Status:  "True",
+					Reason:  "KubeletReady",
+					Message: "kubelet is posting ready status",
+				},
+			},
+		},
+	}
+	node.Name = "ip-172-31-119-125"
+	_, err = c.Nodes().Create(node)
+	require.Nil(suite.T(), err)
+
+	pod := &v1.Pod{
+		Spec: v1.PodSpec{
+			NodeName: node.Name,
+			Containers: []v1.Container{
+				{
+					Name:  "nginx",
+					Image: "nginx:latest",
+				},
+			},
+		},
+	}
+	pod.Name = "nginx"
+	pod.Labels = map[string]string{"app": "nginx"}
+	pendingPod, err := c.Pods("default").Create(pod)
+	require.Nil(suite.T(), err)
+
+	pendingPod.Status = v1.PodStatus{
+		Phase:  "Running",
+		PodIP:  "172.17.0.1",
+		HostIP: "172.31.119.125",
+		Conditions: []v1.PodCondition{
+			{
+				Type:   "Ready",
+				Status: "True",
+			},
+		},
+		// mark it ready
+		ContainerStatuses: []v1.ContainerStatus{
+			{
+				Name:  "nginx",
+				Ready: true,
+				Image: "nginx:latest",
+				State: v1.ContainerState{Running: &v1.ContainerStateRunning{StartedAt: metav1.Now()}},
```

Comment:
How resilient is this to flakiness if there is a clock skew between the API server and the host ?

Reply:
I tried to set a `ContainerStateRunning.StartedAt` in the future (+5 hours) without any failure.
The field returned by the apiserver is untouched: still in the future.

As I played the controller-manager by adding myself the **PodIP NodeName** to the endpoints, the test continue to work as expected.

The metadata.creationTimestamp is accurate and set remotely by the apiserver:
```json
{
  "apiVersion": "v1",
  "kind": "Pod",
  "metadata": {
    "creationTimestamp": "2018-02-09T08:58:48Z",
    "labels": {
      "app": "nginx"
    },
    "name": "nginx",
    "namespace": "default",
    "resourceVersion": "36",
    "selfLink": "/api/v1/namespaces/default/pods/nginx",
    "uid": "710171ce-0d77-11e8-a6ec-000c29dea4f6"
  },
  "spec": {
    "containers": [
      {
        "image": "nginx:latest",
        "imagePullPolicy": "Always",
        "name": "nginx",
        "resources": {},
        "terminationMessagePath": "/dev/termination-log",
        "terminationMessagePolicy": "File"
      }
    ],
    "dnsPolicy": "ClusterFirst",
    "nodeName": "ip-172-31-119-125",
    "restartPolicy": "Always",
    "schedulerName": "default-scheduler",
    "securityContext": {},
    "terminationGracePeriodSeconds": 30
  },
  "status": {
    "conditions": [
      {
        "lastProbeTime": null,
        "lastTransitionTime": null,
        "status": "True",
        "type": "Ready"
      }
    ],
    "containerStatuses": [
      {
        "image": "nginx:latest",
        "imageID": "",
        "lastState": {},
        "name": "nginx",
        "ready": true,
        "restartCount": 0,
        "state": {
          "running": {
            "startedAt": "2018-02-09T13:58:48Z"
          }
        }
      }
    ],
    "hostIP": "172.31.119.125",
    "phase": "Running",
    "podIP": "172.17.0.1"
  }
}
```

Reply:
The **age** reported on kubectl is the **metadata.creationTimestamp**.
The container nginx is Ready: 
```
kubectl -s 127.0.0.1:8080 get po/nginx 
NAME      READY     STATUS    RESTARTS   AGE
nginx     1/1       Running   0          20s
```
The field `startedAt` is probably used for the probes when using an **initialDelaySeconds**.
In our use case, we don't have any kubelet so it's fine.